internal/runtime/registry/registry.go

```diff
@@ -1,55 +1,206 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
 package registry
 
 import (
-	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sync"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
 
 	runtimev1 "sigs.k8s.io/cluster-api/exp/runtime/api/v1beta1"
 	"sigs.k8s.io/cluster-api/internal/runtime/catalog"
 )
 
-func New() Registry {
-	return &registry{
-		registrations: map[string]RuntimeExtensionRegistration{},
-	}
-}
+// TODO: handle namespace selector
+
+var (
+	once     sync.Once
+	instance ExtensionRegistry
+)
+
+// ExtensionRegistry defines all the func an extension registry supports.
+type ExtensionRegistry interface {
+	// WarmUp can be used to initialise a "cold" extension registry with all the known extensions at a given time.
+	// After WarmUp completes the extension registry is considered ready.
+	WarmUp(ext *runtimev1.ExtensionList) error
+
+	// IsReady returns true if the extension registry is ready for usage, and this happens.
+	IsReady() bool
+
+	// Add all the Runtime Extensions served by a registered runtime extension server.
+	Add(ext *runtimev1.Extension) error
+
+	// Remove all the Runtime Extensions served by a registered runtime extension server.
+	Remove(ext *runtimev1.Extension) error
 
-type registry struct {
-	registrations map[string]RuntimeExtensionRegistration
+	// List all the registered Runtime Extensions for a given Runtime Hook.
+	List(gvh catalog.GroupVersionHook) ([]*RuntimeExtensionRegistration, error)
+
+	// Get the Runtime Extensions with a given name
+	Get(name string) (*RuntimeExtensionRegistration, error)
 }
 
+// RuntimeExtensionRegistration contain info about a registered Runtime Extension.
 type RuntimeExtensionRegistration struct {
-	Name             string
+	// The unique name of the Runtime Extension.
+	Name string
+
+	// The GroupVersionHook the Runtime Extension implements.
 	GroupVersionHook catalog.GroupVersionHook
-	ClientConfig     ClientConfig
+
+	ClientConfig   runtimev1.WebhookClientConfig
+	TimeoutSeconds *int32
+	FailurePolicy  *runtimev1.FailurePolicyType
+}
+
+type extensionRegistry struct {
+	ready bool
+	items map[string]*RuntimeExtensionRegistration
+	lock  sync.RWMutex
+}
+
+// Extensions provide access to all the registered Runtime Extensions.
+func Extensions() ExtensionRegistry {
```

Comment:
In general I think it's desirable to only have one registry. I just wonder if we run into issues with this singleton pattern, e.g. for unit/integration testing (outside of the registry package)

Reply:
(but also definitely fine for me to start with that in any case)

Reply:
We can figure it out as soon as we start consuming this from other packages